Tests/KituraWebSocketTests/WebSocketClient.swift

```diff
@@ -135,24 +392,181 @@ class WebSocketMessageHandler: ChannelInboundHandler, RemovableChannelHandler {
+            break
+        }
+    }
+
+    func errorCaught(context: ChannelHandlerContext, error: Error) {
+        if client.delegate != nil {
+            client.delegate?.onError(error: error, status: nil)
+        } else {
+        client.onErrorCallBack(error, nil)
+        }
+        client.disconnect()
+    }
+
+    //  Builds extension headers based on the configuration of maxwindowbits ,context takeover
+    func buildExtensionHeader() -> String {
+        var value = "permessage-deflate"
+        if client.maxWindowBits >= 8 && client.maxWindowBits < 15 {
+            value.append("; " + "client_max_window_bits; server_max_window_bits=" + String(client.maxWindowBits))
+        }
+        value.append(client.contextTakeover.header())
+        return value
+    }
+
+    func upgradeFailure(status: HTTPResponseStatus) {
+        if let delegate = client.delegate {
+            switch status {
+            case .badRequest:
+                delegate.onError(error: WebSocketClientError.badRequest, status: status)
+            case .notFound:
+                delegate.onError(error: WebSocketClientError.webSocketUrlNotRegistered, status: status)
+            default :
+                break
+            }
+        } else {
+            switch status {
+            case .badRequest:
+                client.onErrorCallBack(WebSocketClientError.badRequest, status)
+            case .notFound:
+                client.onErrorCallBack(WebSocketClientError.webSocketUrlNotRegistered, status)
+            default :
+                break
+            }
+        }
+        client.disconnect()
+    }
 }
 
 extension HTTPVersion {
     static let http11 = HTTPVersion(major: 1, minor: 1)
 }
+
+///  This enum is used to populate 'Sec-WebSocket-Extension' field of upgrade header with user required ContextTakeover configuration
+///  User specifies the the context Takeover configuration when creating the WebSocketClient
+///  when not specified both the client and server connections are context takeover enabled
+
+enum ContextTakeover {
+    case none
+    case client
+    case server
+    case both
+
+    func header() -> String {
+        switch self {
+        case .none: return "; client_no_context_takeover; server_no_context_takeover"
+        case .client: return "; server_no_context_takeover"
+        case .server: return "; client_no_context_takeover"
+        case .both: return ""
+        }
+    }
+
+    var clientNoContextTakeover: Bool {
+        return self != .client && self != .both
+    }
+
+    var serverNoContextTakeover: Bool {
+        return self != .server && self != .both
+    }
+}
+
+enum WebSocketClientError: UInt, Error {
+    case webSocketUrlNotRegistered = 404
+    case badRequest = 400
+
+    func code() -> UInt? {
+        switch self {
+        case .webSocketUrlNotRegistered :
+            return 404
+        case .badRequest :
+            return 400
+        }
+    }
+}
+
+public protocol WebSocketClientDelegate {
```

Comment:
Please remove this protocol if we aren't using it in the tests. We can include it when we create a new package for the WebSocket client.